### PR TITLE
New Feature - PolygonSelector Widget

### PR DESCRIPTION
**Pull request by bduick**
Hello matplotlib community! Please checkout the PolygonSelector widget.  The demo is pretty straight forward, and basically a copy of the LassoSelector demo.  I'm open to any feedback, or suggestions for improvement.  Thanks for your consideration.

demo-screen-captures

### CODE REVIEW

File: lib/matplotlib/tests/test_widgets.py [hunk 1]
```diff
@@ -313,3 +313,126 @@ def test_slider_valmin_valmax():
...
+    expected_result = [(50, 50), (150, 50), (50, 150)]
+    event_sequence = [*polygon_place_vertex(50, 50),
+                      *polygon_place_vertex(150, 50),
+                      *polygon_place_vertex(50, 150),
+                      *polygon_place_vertex(50, 50)]
+    check_polygon_selector(event_sequence, expected_result, 1)
+
+    # Move first vertex before completing the polygon.
+    expected_result = [(75, 50), (150, 50), (50, 150)]
+    event_sequence = [*polygon_place_vertex(50, 50),
```
Comment (phobson) [inline lib/matplotlib/tests/test_widgets.py]: unfortunately, we have to support python 2.7 and 3.4, so this won't work :/

File: lib/matplotlib/tests/test_widgets.py [hunk 2]
```diff
@@ -313,3 +313,126 @@ def test_slider_valmin_valmax():
...
+        The list of vertices that are expected to result from the event
+        sequence.
+    selections_cnt : int
+        Wait for the tool to call its `onselect` function `selections_cnt`
+        times, before comparing the result to the `expected_result`
+    """
+    ax = get_ax()
+
+    ax._onselect_cnt = 0
+    def onselect(vertices):
```
Comment (phobson) [inline lib/matplotlib/tests/test_widgets.py]: PEP8 wants a blank like here

**Comment by bduick** [comment]
Thanks phobson. Glad you like it.

File: lib/matplotlib/widgets.py [hunk 3]
```diff
@@ -2499,6 +2499,217 @@ def _onmove(self, event):
...
+
+    Parameters
+    ----------
+    ax : :class:`~matplotlib.axes.Axes`
+        The parent axes for the widget.
+    onselect : function
+        When a polygon is completed or modified after completion,
+        the `onselect` function is called and passed a list of the vertices as
+        ``(xdata, ydata)`` tuples.
+    useblit : bool, optional
```
Comment (tacaswell) [inline lib/matplotlib/widgets.py]: Interactive is not documented.

File: lib/matplotlib/widgets.py [hunk 4]
```diff
@@ -2499,6 +2499,217 @@ def _onmove(self, event):
...
+    lineprops : dict, optional
+        The line for the sides of the polygon is drawn with `lineprops`. The
+        default is ``dict(color='k', linestyle='-', linewidth=2, alpha=0.5)``.
+    markerprops : dict, optional
+        The markers for the vertices of the polygon are drawn with
+        `markerprops`. The default is ``dict(marker='o', markersize=7, mec='k',
+        mfc='k', alpha=0.5)``.
+    vertex_select_radius : int
```
Comment (tacaswell) [inline lib/matplotlib/widgets.py]: should be optional

**Comment by tacaswell** [comment]
This looks useful :+1: 

attn @blink1073 I have completely lost track of where you are at on other selector work.

**Comment by bduick** [comment]
@tacaswell i can add support for non-interactive. i think i will implement it so that the only difference with interactive mode is that the polygon is automatically cleared after the selection is made.  alternatively, does anyone think that i should also inhibit the _move_vertex_ and _move_all_ functions that are currently allowed before the polygon has been completed? in other words, the way it currently works is that if you haven't closed the polygon yet, you can move all the vertices or move a single vertex by holding _shift_ or _ctrl_. should i disable this in non-interactive to be more consistent with the other widgets?

**Comment by tacaswell** [comment]
On a second reading, just drop `interactive` as a kwarg.

**Comment by blink1073** [comment]
This is great!  I had tried and given up on a similar tool in another PR.

**Comment by bduick** [comment]
My bad on the `interactive` kwarg.  Didn't even realize until after your latest comment that I had accidentally left it in the signature.

**Comment by bduick** [comment]
@blink1073, glad you like the tool. i found your [PR](https://github.com/matplotlib/matplotlib/issues/5782) after thomas tagged you. the PR makes a lot of good points. i did find that i had to put in some special logic to work around logic in `_SelectorWidget` that wasn't applicable for the `PolygonSelector`.

File: lib/matplotlib/widgets.py (hunk 4 above)
Comment (tacaswell) [inline lib/matplotlib/widgets.py]: and would this be better as a float?

**Comment by tacaswell** [comment]
@bduick Ah, sorry my comments were cryptic.  Could you add to the docstring that `'vertex_select_radius'` is optional?

I think that is the only thing standing between this and merging :)

File: examples/widgets/polygon_selector_demo.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+from matplotlib.widgets import PolygonSelector
+from matplotlib.path import Path
+
+
+class SelectFromCollection(object):
+    """Select indices from a matplotlib collection using `PolygonSelector`.
+
+    Selected indices are saved in the `ind` attribute. This tool highlights
+    selected points by fading them out (i.e., reducing their alpha values).
```
Comment (QuLogic) [inline examples/widgets/polygon_selector_demo.py]: Doesn't it fade out the points that are *not* selected?

**Comment by bduick** [comment]
I should've understood. Anyways, fixed and committed!

File: examples/widgets/polygon_selector_demo.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+        self.xys = collection.get_offsets()
+        self.Npts = len(self.xys)
+
+        # Ensure that we have separate colors for each object
+        self.fc = collection.get_facecolors()
+        if len(self.fc) == 0:
+            raise ValueError('Collection must have a facecolor')
+        elif len(self.fc) == 1:
+            self.fc = np.tile(self.fc, self.Npts).reshape(self.Npts, -1)
```
Comment (QuLogic) [inline examples/widgets/polygon_selector_demo.py]: Isn't this `np.repeat(self.fc, self.Npts)`?

File: examples/widgets/polygon_selector_demo.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+        self.poly = PolygonSelector(ax, self.onselect)
+        self.ind = []
+
+    def onselect(self, verts):
+        path = Path(verts)
+        self.ind = np.nonzero([path.contains_point(xy) for xy in self.xys])[0]
```
Comment (QuLogic) [inline examples/widgets/polygon_selector_demo.py]: Does `path.contains_points(self.xys)` work here?

File: lib/matplotlib/tests/test_widgets.py [hunk 8]
```diff
@@ -313,3 +313,139 @@ def test_slider_valmin_valmax():
...
+    """Helper function to test Polygon Selector
+
+    Parameters
+    ----------
+    event_sequence : list of tuples (etype, dict())
+        A sequence of events to perform. The sequence is a list of tuples
+        where the first element of the tuple is an etype (e.g., 'onmove',
+        'press', etc.), and the second element of the tuple is a dictionary of
+         the arguments for the event (e.g., xdata=5, key='shift', etc.).
+    expected_result : list or vertices (xdata, ydata)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_widgets.py]: list of?

File: lib/matplotlib/tests/test_widgets.py [hunk 9]
```diff
@@ -313,3 +313,139 @@ def test_slider_valmin_valmax():
     slider = widgets.Slider(ax=ax, label='', valmin=0.0, valmax=24.0,
                             valinit=25.0)
     assert slider.val == slider.valmax
+
+
+def check_polygon_selector(event_sequence, expected_result, selections_cnt):
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_widgets.py]: Don't see any reason to abbreviate `count`.

File: lib/matplotlib/tests/test_widgets.py [hunk 10]
```diff
@@ -313,3 +313,139 @@ def test_slider_valmin_valmax():
...
+    check_polygon_selector(event_sequence, expected_result, 1)
+
+    # Move first vertex after completing the polygon.
+    expected_result = [(75, 50), (150, 50), (50, 150)]
+    event_sequence = (polygon_place_vertex(50, 50)
+                      + polygon_place_vertex(150, 50)
+                      + polygon_place_vertex(50, 150)
+                      + polygon_place_vertex(50, 50)
+                      + [('onmove', dict(xdata=50, ydata=50)),
+                        ('press', dict(xdata=50, ydata=50)),
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_widgets.py]: Misaligned.

File: lib/matplotlib/widgets.py [hunk 11]
```diff
@@ -2499,6 +2499,217 @@ def _onmove(self, event):
         self.update()
 
 
+class PolygonSelector(_SelectorWidget):
+    """Select a polygon region of an axes.
+
+    Place vertices with each mouse click, and make the selection by completing
+    the polygon (clicking on the first vertex). Hold the *ctrl* key and click
+    and drag a vertex to reposition it (the *ctrl* key is not necessary if the
+    polygon has already be completed). Hold the *shift* key and click and drag
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: already been

File: lib/matplotlib/widgets.py [hunk 12]
```diff
@@ -2499,6 +2499,217 @@ def _onmove(self, event):
...
+    ----------
+    ax : :class:`~matplotlib.axes.Axes`
+        The parent axes for the widget.
+    onselect : function
+        When a polygon is completed or modified after completion,
+        the `onselect` function is called and passed a list of the vertices as
+        ``(xdata, ydata)`` tuples.
+    useblit : bool, optional
+    lineprops : dict, optional
+        The line for the sides of the polygon is drawn with `lineprops`. The
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: "... drawn with the properties given by ``lineprops``."

(Also, references do not work to arguments.)

File: lib/matplotlib/widgets.py [hunk 13]
```diff
@@ -2499,6 +2499,217 @@ def _onmove(self, event):
...
+        default is ``dict(color='k', linestyle='-', linewidth=2, alpha=0.5)``.
+    markerprops : dict, optional
+        The markers for the vertices of the polygon are drawn with
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: "... drawn with the properties given by ..."

File: lib/matplotlib/widgets.py [hunk 14]
```diff
@@ -2499,6 +2499,217 @@ def _onmove(self, event):
...
+        self.vertex_select_radius = vertex_select_radius
+
+        self.artists = [self.line, self._polygon_handles.artist]
+        self.set_visible(True)
+
+    def _press(self, event):
+        """Button press event handler"""
+        # Check for selection of a tool handle.
+        if ((self._polygon_completed or 'move_vertex' in self.state)
+            and len(self._xs) > 0):
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: Should be indented one more time.

File: lib/matplotlib/widgets.py [hunk 15]
```diff
@@ -2499,6 +2499,217 @@ def _onmove(self, event):
...
+                and ('move_vertex' in self.state or 'move_all' in self.state)):
+            self._xs, self._ys = self._xs[:-1], self._ys[:-1]
+            self._draw_polygon()
+
+    def _on_key_release(self, event):
+        """Key release event handler"""
+        # Add back the pending vertex if leaving the 'move_vertex' or
+        # 'move_all' mode (by checking the released key)
+        if (not self._polygon_completed
+            and (event.key == self.state_modifier_keys.get('move_vertex')
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: Should be indented once more.

File: lib/matplotlib/widgets.py [hunk 16]
```diff
@@ -2499,6 +2499,217 @@ def _onmove(self, event):
...
+            self.set_visible(True)
+
+    def _draw_polygon(self):
+        """Redraw the polygon based on the new vertex positions."""
+        self.line.set_data(self._xs, self._ys)
+        # Only show one tool handle at the start and end vertex of the polygon
+        # if the polygon is completed or the user is locked on to the start
+        # vertex.
+        if (self._polygon_completed
+            or (len(self._xs) > 3
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py]: Should be indented once more.

File: examples/widgets/polygon_selector_demo.py (hunk 6 above)
Comment (QuLogic) [inline examples/widgets/polygon_selector_demo.py]: Actually, no, I just tried it out and it's not, but it is equal to `np.tile(x, (20, 1))` if I've understood it correctly.

**Comment by bduick** [comment]
@QuLogic, thanks for the thorough review.  Your 3 comments on the polygon_selector_demo.py also apply to [lasso_selector_demo.py](http://matplotlib.org/examples/widgets/lasso_selector_demo.html).  Can I commit the fixes for the lasso demo in this PR, or should I create a new one?

File: lib/matplotlib/widgets.py (hunk 12 above)
Comment (bduick) [inline lib/matplotlib/widgets.py]: Can you give me some more info re: "references do not work to arguments"? I'm not sure what you mean.

**Comment by tacaswell** [comment]
@bduick which ever is easier for you.

**Comment by tacaswell** [comment]
failures look to be due to changing DNS for matplotlib.org and it taking time for those changes to propogate

**Comment by ordirules** [comment]
very cool!

File: lib/matplotlib/widgets.py (hunk 12 above)
Comment (bduick) [inline lib/matplotlib/widgets.py]: @QuLogic, is there anything else I need to do to address this comment?

**Comment by tacaswell** [comment]
attn @phobson

**Pull request merged**